crates/extensions/tedge_mqtt_bridge/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+            msgs_cloud,
+            Some(health_topic),
+            "cloud",
+        ));
+
+        Self {}
+    }
+
+    pub(crate) fn build_actor(self) -> MqttBridgeActor {
+        MqttBridgeActor {}
+    }
+}
+
+fn bidirectional_channel<T>(buffer: usize) -> [BidirectionalChannelHalf<T>; 2] {
+    let (tx_first, rx_first) = mpsc::channel(buffer);
+    let (tx_second, rx_second) = mpsc::channel(buffer);
+    [
+        BidirectionalChannelHalf {
+            tx: tx_first,
+            rx: rx_second,
+        },
+        BidirectionalChannelHalf {
+            tx: tx_second,
+            rx: rx_first,
+        },
+    ]
+}
+
+struct BidirectionalChannelHalf<T> {
+    tx: mpsc::Sender<T>,
+    rx: mpsc::Receiver<T>,
+}
+
+impl<'a, T> BidirectionalChannelHalf<T> {
+    pub fn send(&'a mut self, item: T) -> futures::sink::Send<'a, mpsc::Sender<T>, T> {
+        self.tx.send(item)
+    }
+
+    pub fn recv(&mut self) -> futures::stream::Next<mpsc::Receiver<T>> {
+        self.rx.next()
+    }
+}
+
+/// Forward messages received from `recv_event_loop` to `target`
+///
+/// The result of running this function constitutes half the MQTT bridge, hence the name.
+/// Each half has two main responsibilities, one is to take messages received on the event
+/// loop and forward them to the target client, the other is to communicate with the corresponding
+/// half of the bridge to ensure published messages get acknowledged only when they have been
+/// fully processed.
+///
+/// # Message flow
+/// Messages in the bridge go through a few states
+/// 1. Received from the sending broker
+/// 2. Forwarded to the receiving broker
+/// 3. The receiving broker sends an acknowledgement
+/// 4. The original forwarded message is acknowledged now we know it is fully processed
+///
+/// Since the function is processing one [EventLoop], messages can be sent to the receiving broker,
+/// but we cannot receive acknowledgements from that broker, therefore a communication link must
+/// be established between the bridge halves. This link is the argument `corresponding_bridge_half`,
+/// which can both send and receive messages from the other bridge loop.
+///
+/// When a message is forwarded, the [Publish] is forwarded from this loop to the corresponding loop.
+/// This allows the loop to store the message along with its packet ID when the forwarded message is
+/// published. When an acknowledgement is received for the forwarded message, the packet id is used
+/// to retrieve the original [Publish], which is then passed to [AsyncClient::ack] to complete the
+/// final step of the message flow.
+///
+/// The channel sends [Option<Publish>] rather than [Publish] to allow the bridge to send entirely
+/// novel messages, and not just forwarded ones, as attaching packet IDs relies on pairing every
+/// [Outgoing] publish notification with a message sent by the relevant client. This means the
+///
+/// # Health topics
```

Comment:
> /// The channel sends [Option<Publish>] rather than [Publish] to allow the bridge to send entirely
/// novel messages, and not just forwarded ones, as attaching packet IDs relies on pairing every
/// [Outgoing] publish notification with a message sent by the relevant client. This means the

I don't get it. What is the point to send a `None` to the half? Or to send *entirely novel* message? Even is useful, how `Option` can help?

> This means the

missing sentence

> # Health topics

missing content

Reply:
Hopefully all is clearer now I've finished the paragraph and the next section